### PR TITLE
Fullscreen Mode: Change W button to toggle wp-admin sidebar

### LINKED ISSUE #22178: Usability: Full screen editor makes side and top bar hard to access
The default full screen mode hides so much navigation behind extra clicks as to make it less intuitive. The interaction for "new post" as well as "view post" each are difficult to find, for both long time and new users alike, and could be displayed on the edit screen without the need for extra clicks. [Please see this video demonstration](https://www.loom.com/share/6ae9a43d049b43e7a788129dff66bfec). 

Having been a WordPress user for over a decade, it took me 10 minutes of discussion with a colleague to determine that I could make these interactions more accessible by exiting fullscreen mode - and even then, fullscreen toggling requires extra clicks to find. In the pre-gutenberg editor, viewing a post, or adding a new post from the edit screen could be accomplished in one click. In the current version this requires a minimum of two clicks, and depending on the approach used, can require navigating to the dashboard from the edit page, requiring a page load. This makes the user experience less intuitive and more challenging, especially for users new to WordPress or to the new editor, but for long-time users as well. 

- WordPress version: 5.4.1
- Default installation

All OS, Browsers, Devices etc.

### PR DESCRIPTION
**Pull request by shaunandrews**
Fixes #22178

This is a second attempt; The first (and a lot more background info) is #21121.

> With the recent move to make Fullscreen mode the default, I've heard from many people that they expect to be able to toggle the WordPress sidebar menu by clicking on the WordPress logo in the top left. This PR tries that:

wp-admin-toggle

### CODE REVIEW

**Comment by MichaelArestad** [comment]
Using this feels solid. I only have minor feedback:

In Chromium (Chrome and Edge Mac), I noticed there is an outline style around the W when clicked. It does not show up in Firefox. It made me think I had to click within the blue border to close it.

image

The other minor weirdness is around the speed of it closing. When I click in to the editor, it closes automatically, which is great. It does feel like it takes a touch too long. The timing might be good to explore in a future issue/PR.

**Comment by talldan** [comment]
I couldn't see a linked issue, and #22178 was reported, so I've linked this PR as closing that issue.

**Comment by youknowriad** [comment]
Seems like there's  a coupe bugs here

Capture d’écran 2020-05-08 à 3 52 45 PM

The safari issue is back (white spacec)

Capture d’écran 2020-05-08 à 3 52 55 PM

I think this state was one of the reasons the revert was suggested, should we hide the "Collapse menu" button on Fullscreen?

Capture d’écran 2020-05-08 à 3 53 06 PM

Also from that state, when you click the "w" again, you get this.

---

I think one of the other reasons, the revert was suggested was whether the "W" should be centered or not when the panel is open and potentially add the site title there? cc @mtias

**Comment by shaunandrews** [comment]
Thanks @youknowriad — some of these I knew, some I didn't.

> I think one of the other reasons, the revert was suggested was whether the "W" should be centered or not when the panel is open and potentially add the site title there?

I don't have a strong preference on the centering, but I also haven't heard it brought up before. 

Adding a site title was explored, but I still think its best left for another PR so we can figure out a good way to displaying titles without aggressive truncation.

There was mention of having to integrate the wp-admin color schemes to the W button, and the lack of a selected state for the button. I'm still working on both of those.

**Comment by jasmussen** [comment]
Thanks for your endurance, Shaun.

It feels very important to me that users can _middle-click to open in a new tab_ on the logo. For people who do this, they expect to be able to, and when they can't, it can erode a bit of trust. In the before-times, we would keep the `<a href="...">` intact, to make this work, but add an `onclick` handler with a `return false` to stop the URL navigation.

The focus style doesn't meet contrast ratio. You could either do the double-shot version we have in master, or change the focus style to be white there. 

Screenshot 2020-05-11 at 09 16 19

Screenshot 2020-05-11 at 09 17 15

I'm happy to see you were able to get the sidebar scrolling on short viewports, or admins with a lot of menu items:

Screenshot 2020-05-11 at 09 18 29

It's unfortunate that we still have to have the double-scrollbar on the right to make this happen; the fact that we got rid of that double scrollbar with fullscreen was my personal favorite feature of that. But at least now it's consistent with non-fullscreen. If your neurons can think up a way to make the left navigation rail scroll independantly, as a separate PR to this, I would buy you a milkshake. No, make that two milkshakes.

**Comment by shaunandrews** [comment]
> It feels very important to me that users can middle-click to open in a new tab on the logo.

I haven't forgotten about this; Next on my list here.

> The focus style doesn't meet contrast ratio. You could either do the double-shot version we have in master, or change the focus style to be white there.

I added a second, white border similar to what `master` does.

> It's unfortunate that we still have to have the double-scrollbar on the right to make this happen; the fact that we got rid of that double scrollbar with fullscreen was my personal favorite feature of that. 

I pushed a (hopeful) fix for this; Basically, I forced `overflow: hidden;` when full screen mode is enabled, and the sidebar is hidden. This feels like a better experience, in that you'd only see the double-scrollbar when you've toggled the sidebar. Let me know what you think, @jasmussen

--

I managed to update the colors to respect the wp-admin color scheme settings, which required adding two new values to the `base-styles` package.

Along with the request to allow a middle-click, I'm still considering wether the "folded" or "collapsed" mode of the sidebar makes sense in this context.

**Comment by jasmussen** [comment]
> I added a second, white border similar to what master does.

I think we should try just the white, this feels a little heavy:

focus

There's also a `$border-width-focus` you can use so it's consistent with all other focus, instead of the 4px picked here:

Screenshot 2020-05-14 at 10 02 15

I also wonder whether it's actually nice for the focus to encompass the entire button rather than try to frame just the logo. Like so:

focus alt

That should translate to this:

```
    box-shadow: inset 0 0 0 $border-width-focus #23282e, inset 0 0 0 #{ $border-width-focus + $border-width } $white;
```

I noticed a weird focus trap of sorts, though:

focus trap

In short, the tabbing context changes depending on whether the menu is open or not. If the menu is closed, you tab the top menu. If the menu is open, you tab through the menu. Both these seem fine, the problem is I can't shift tab backwards from the menu into the logo anymore. Not sure what the best practice is here. 

> I pushed a (hopeful) fix for this; Basically, I forced overflow: hidden; when full screen mode is enabled, and the sidebar is hidden. This feels like a better experience, in that you'd only see the double-scrollbar when you've toggled the sidebar. Let me know what you think, @jasmussen

scrollbars

This seems to work as intended!

Still not ideal, but I don't think it can ever become ideal until we are able to touch the left navigation side rail directly.

And that, I think, actually the biggest challenge here — it's bridging old and new, and the edgecases that pop up, including what you mention about folded cases and such, are really really difficult to address in a good way. I'm impressed that you're tackling it head-on!

**Comment by ItsJonQ** [comment]
👋 Haiii!! I can help out with the JS/interaction side of things :)

@jasmussen Feedback/questions for you!

> It feels very important to me that users can middle-click to open in a new tab on the logo...

Gotcha. So allow for the W logo to work like a link.
It's currently a `<button />` element. We can change it to a link.
If we do, what should the URL be?

For example, when I edit a post, the URL is:

```
http://localhost:8889/wp-admin/post.php?post=1080&action=edit
```

> ...I can't shift tab backwards from the menu into the logo anymore...

We can accommodate this. Just a heads up, the solution will be hacky. It's already quite hacky, but it'll be hackier. There isn't a way around this as the (React) `AdminMenuToggle` component is doing things in non-React `wp-admin` land.

Thanks!

**Comment by jasmussen** [comment]
> If we do, what should the URL be?

I would use the same heuristic as we have today, it has felt natural for me for the duration of using fullscreen. I'm not 100% on what that heuristic is, I think it's posts when you edit posts, pages when you edit pages, and probably cpts work as well. So to an extent, I think of this more as "restoring the current behavior".

> We can accommodate this. Just a heads up, the solution will be hacky. It's already quite hacky, but it'll be hackier. There isn't a way around this as the (React) AdminMenuToggle component is doing things in non-React wp-admin land.

Let's be liberal with code comments, and let's also take notes for how we could do this better. I hope that we'll sooner rather than later be able to revisit the actual side navigation rail to improve things there, and if/when we do it would be good to know what to refactor to make things better.

File: packages/edit-site/src/components/editor/index.js [hunk 1]
```diff
@@ -92,6 +97,9 @@ function Editor( { settings: _settings } ) {
 		<>
 			<EditorStyles styles={ settings.styles } />
 			<FullscreenMode isActive={ isFullscreenActive } />
+			<MainDashboardButton.Slot>
```
Comment (vindl) [inline packages/edit-site/src/components/editor/index.js:100]: I tried testing this with instructions from https://github.com/WordPress/gutenberg/pull/22323 to verify if extensibility of the close button, and it doesn't seem to work anymore. The default button is not replaced.

I'm not sure what's causing this, and whether it's somehow related to moving this outside of editor skeleton. FWIW I don't think that it should be moved out of this area. With this setup it seems that everyone who's aiming to provide a custom button, will also have to provide styles to position it properly on top of the header, and I think that will be prone to breakage over time.

**Comment by ItsJonQ** [comment]
@jasmussen + @shaunandrews Haiii!

I just pushed some updates that handles the 2 enhanced interactions:

1. Middle clicking (or ctrl/cmd + click) the (W) logo to open the in a new tab
2. Pressing <kbd>Shift</kbd>+<kbd>Tab</kbd> on the first WP Admin menu item to jump focus back to the Gutenberg toolbar (W) logo.

I've added a bunch of comments that describe how it works, and acknowledging that it's perhaps not the best 😅 

Let me know how it feels!

**Comment by ItsJonQ** [comment]
Screen Shot 2020-05-19 at 1 18 54 PM

Oo! Pushed another update. This one updates the (Tooltip) label of the AdminMenuToggle (W) link to say "Show" / "Hide" depending on the open state.

Previously, it always said "Show"

**Comment by shaunandrews** [comment]
@ItsJonQ Love the updates; Thanks!

My only bit of feedback is that the `href` should point to `edit.php` with `addQueryArgs` for the current document's `post_type`, like the old close button used to do:

https://github.com/WordPress/gutenberg/pull/22191/files#diff-5e54b8ed48d99cb46a591a186b9ad138L37

I tried this locally, but for some reason I'm getting `undefined` for the `postType.slug`.

**Comment by ItsJonQ** [comment]
@shaunandrews Thank you! I'm happy to change the URL of the `href`. I couldn't find it from the link you shared. Would you happen to have the code example to get that `edit.php` based URL from the old Button?

Thank you!

**Comment by shaunandrews** [comment]
Sorry, the link highlights a deleted file. Its strange. I copy/pasted the old button code:

```function FullscreenModeClose() {
 	const { isActive, postType } = useSelect( ( select ) => {
 		const { getCurrentPostType } = select( 'core/editor' );
 		const { isFeatureActive } = select( 'core/edit-post' );
 		const { getPostType } = select( 'core' );

 		return {
 			isActive: isFeatureActive( 'fullscreenMode' ),
 			postType: getPostType( getCurrentPostType() ),
 		};
 	}, [] );

 	if ( ! isActive || ! postType ) {
 		return null;
 	}

 	return (
 		<Button
 			className="edit-post-fullscreen-mode-close"
 			icon={ wordpress }
 			iconSize={ 36 }
 			href={ addQueryArgs( 'edit.php', {
 				post_type: postType.slug,
 			} ) }
 			label={ get( postType, [ 'labels', 'view_items' ], __( 'Back' ) ) }
 		/>
 	);
 }```

**Comment by ItsJonQ** [comment]
@shaunandrews Awesome! Thank you! I'll give it a shot

**Comment by ItsJonQ** [comment]
@shaunandrews I just pushed up a change. It looks like it's working for me 😊 . Would you be able to let me know?

(Also. I'll fix up that merge conflict)

**Comment by shaunandrews** [comment]
> It looks like it's working for me

That's cause you did it a way better way than my failed attempt.

Something changed with the `tab` behavior; I can no longer use the keyboard to `tab` into the sidebar menu.

**Comment by ItsJonQ** [comment]
> Something changed with the tab behavior; I can no longer use the keyboard to tab into the sidebar menu.

Ah yes, my bad! I broke the HTML element reference. I just pushed that fix, along with rebase with `master`.

**Comment by paaljoachim** [comment]
Hey @shaunandrews and others can we get an update to how this PR is coming along.
Thanks.

Can we look at the possibility of getting this and associated issue/s into WordPress 5.5?

**Comment by ItsJonQ** [comment]
Hmm! Looks like it has conflicts again. Will help resolve

File: packages/interface/src/components/admin-menu-toggle/index.js [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+		/**
+		 * Otherwise, toggle the admin menu.
+		 */
+		event.preventDefault();
+		toggleMenu();
+	};
+
+	const label = isOpen
+		? __( 'Hide sidebar menu' )
```
Comment (moorscode) [inline packages/interface/src/components/admin-menu-toggle/index.js:37]: Technically this is a "sidebar menu", but isn't this the "admin navigation" of a users-site?
"Hide navigation" or "Hide admin navigation" would make more sense to me as a (non-technical) user.

Comment (ItsJonQ) [reply packages/interface/src/components/admin-menu-toggle/index.js:37]: Thanks for the feedback! I'll make that update

**Comment by mtias** [comment]
I want to sit on this one a bit more before we commit to adding it in time for 5.5. The main issue I see is in the case we are looking for the W menu to open page list / template list in the context of FSE or the page editor, and how it might end up being disruptive if we change the behaviour version to version.

**Comment by shaunandrews** [comment]
Closing this for now.

**Pull request closed without merging**